README
SP-ENDURANCE INSTRUCTIONS
=========================

Collecting Endurance Snapshots
------------------------------

- Pre-requirements:
  - Kernel with SMAPS enabled and normal POSIX tools
    (like Maemo and about all normal Linux distributions)
  - sp-endurance package and its dependencies installed

1. If you don't want to have lot of uninteresting log data,
   clean the syslog (as root), e.g. with:
     echo "" > /var/log/syslog.old
     echo "" > /var/log/syslog
   (If your logs are in normal place)
2. Reboot the device
3. If you have X, DISPLAY is not set and needs to be something else
   than ":0", set it (its needed for xmeminfo run by the script):
     export DISPLAY=:1
4. Then run the following before first test round:
     endurance-snapshot usecase
  -> This will save data files under "usecase/XXX/" directory.
     Each time it's called, XXX increases (000, 001, 002, ...)
5. Repeat step 4) after each test round.
   Note: if you're going to do this many times for use-cases which
   write a lot of data to syslog, you might clean the syslog also
   after every step 4)

Note: The purpose of the first snapshot in step 4) is to have separate
values for state before tests and how much resources are used when
everything involved in the test round are initialized after that. 
Rest of the steps where the same use-case is repeated, are then about
catching resource leakage (e.g. inappropriate caching of resources).


With the "endurance-mem-overview" script one can quickly check from
the collected data whether there were (already) large differencies
in the whole system memory usage.  This way only individual process
leakage and the leakage of other (less frequently leaked) resources
needs to be investigated with the full post-processing tools.


Post-processing / visualizing the endurance data
------------------------------------------------

- Pre-requirements:
  - sp-endurance-postproc package and its dependencies installed
  - the contents of the "usecase" directory (collected endurance data)

1. Run:
     endurance-parse-snapshots usecase/[0-9]*
2. Run:
     endurance-plot usecase/[0-9]*

Step 1 produces:
- HTML endurance report to file "endurance-report.html". This file is
  self-contained, and can be easily attached to bug reports, emails, etc.
  The data is shown in numeric and bar graph formats.
- The syslogs from each endurance is processed for interesting entries, such as
  error or warning messages from the kernel or user space programs.
  endurance-report.html automatically contains links to each of these reports.

Step 2 produces:
- Endurance report in graphical format, that can be easily browsed by opening
  the resulting "index.html" file.

If you want also details memory usage information from the
applications, give additional --parse-smaps argument to the first
script. This will require sp-smaps-visualize package. Or just read
the smaps.cap.gz files directly (endurance-extract-process-smaps
extracts just the given process information from all rounds).


Files coming from the source package
------------------------------------

Files for the endurance data collection:
- measure/endurance-snapshot
  - saves one snapshot of endurance data under the given directory
    using the tools below, date, df, sp_smaps_snapshot etc
- measure/proc2csv
  - binary required by endurance-snapshot
  - outputs contents of /proc/ files as CSV data
- measure/xmeminfo
  - binary optionally used by endurance-snapshot
  - outputs X clients resource usage from the X server as CSV data
- endurance-mem-overview
  - show ASCII chart of system memory usage from the already
    collected endurance data

Files for the data post-processing:
- postproc/endurance-parse-snapshots
  - runs endurance-report with suitable options
- postproc/endurance-plot
  - creates a large variety of (RAM, swap, CPU, X resource usage, interrupt,
    context switch etc) graphs from endurance data collected with
    endurance-snapshot for trend analysis
- postproc/endurance-report
  - post-processes data files saved by endurance-snapshot
- postproc/syslog_parse.py
  - used by endurance-report
  - script for digging out known types of problems from syslogs
- postproc/endurance-split-snapshots
  - script for making several reports out of given data files
    (e.g. if device has booted in middle of testing, that invalidates
    the memory changes, so you want a separate report after that)
- postproc/endurance-extract-process-smaps
  - script for extracting just the given process SMAPS information
    from all the endurance rounds for comparising (diff -u ...)
- postproc/endurance-extract-process-cgroup
  - script for extracting Cgroup process information from endurance data.
  - generates output in either text or HTML format.

Other files:
- README
  - this file
- src/proc2csv.c
  - source code for proc2csv
- src/xmeminfo.c
  - source code for xmeminfo
- Makefile
  - compiles the above source files and installs them
- man/
  - manual pages for all binaries and scripts
- postproc-lib/
  - perl module required by some of the post processing tools


changelog
sp-endurance (4.0.0), 8 Oct 2012:
 * Program names have been cleaned up:
   - "save-incremental-endurance-stats"  => "endurance-snapshot"
   - "endurance_plot"                    => "endurance-plot"
   - "endurance_report.py"               => "endurance-report"
   - "extract-endurance-process-cgroups" => "endurance-extract-process-cgroups"
   - "extract-endurance-process-smaps"   => "endurance-extract-process-smaps"
   - "parse-endurance-measurements"      => "endurance-parse-snapshots"
   - "recompress-endurance-measurements" => "endurance-recompress-snapshots"
   - "split-endurance-measurements"      => "endurance-split-snapshots"
   - The old names may still be still made available by installing symbolic
     links with "make install-compat-symlinks".
 * Snapshot directory numbering has been cleaned up:
   - The old numbering 101, 102, 103, ... is replaced with 000, 001, 002, ...
 * endurance-snapshot:
   - Collect Upstart v1.4+ service logs from /var/log/upstart/.
   - Collect /proc/pid/sched.
   - Avoid collecting empty files.
 * The "endurance-plot" script has been cleaned up and improved:
   - The parsing and plotting logic has been rewritten and split into a
     reusable and unit tested Perl module.
   - The /proc/pid/smaps parsing has been implemented with C to provide
     speedup compared to the pure Perl implementation. This feature is
     optional, and requires that the "Inline" Perl module is available at
     build time.
   - Per process I/O graphs have been added.
   - Private Dirty graphs have been added.
   - Upstart "Jobs Respawned" graph has been added.
   - Process scheduling graphs have been added.
   - JSON output format in addition to HTML.
 * endurance-report:
   - Tolerate missing data and various I/O errors better.
   - Performance improvements in the syslog parser.
 * endurance-recompress-snapshots:
   - The level of parallellism is now fully user controlled via '-j N' flag.
 * endurance-multiplot:
   - A new experimental endurance graphing script for multiple usecases.

sp-endurance (3.0.4), 23 Mar 2012:
 * parse IO disk write stats from usage.csv to endurance report.

sp-endurance (3.0.3), 27 Feb 2012:
 * proc2csv fixes for newer kernels (desktop Ubuntu):
   - change ',' in proc files field contents to '/'
   - increase proc file CSV line buffer size from 256 to 512,
     otherwise long lines can get split
 * support separate doc install path
 * endurance_plot:
   - Fix incorrect entries for short living processes in some histogram plots.
   - 'Major page fault' graphs now split to several graphs for improved
     readability. Graphs added for minor page faults.
   - Added 'private dirty' memory graphs.
   - Virtual memory size graph now split to several graphs for improved
     readability.

sp-endurance (3.0.2), 27 Dec 2011:
 * Harmattan 'sp-endurance-tests' package integration fix.

sp-endurance (3.0.1), 22 Dec 2011:
 * endurance_plot:
   - Rename '#smaps segments' plot to '#memory maps'.
   - Split the kernel slab graph to count and changes graphs.
   - Split the single ramzswap graph into three separate graphs:
     1. Reads and writes in megabytes per round.
     2. Memory usage: OrigDataSize, ComprDataSize, MemUsedTotal, ZeroPages,
        GoodCompress and NoCompress.
     3. Errors, if any (in most cases this graph will not be created).
   - Fix wchan parsing /proc/pid/io data from usage.csv.
   - Add backlight brightness information to battery graph.
 * endurance_report.py:
   - Do not warn about missing data for 'sp-noncached'.
   - Add hardware information to the endurance report by parsing the
     `component_version' file from the endurance data.
   - Add battery information table.

sp-endurance (3.0), 9 Dec 2011:
 * NOTE: this release introduces changes to the collected endurance data.
   Post-processing tools are able to process both new and old formats.
 * X support/dependency is optional in build, measuring and post-processing.
 * save-incremental-endurance-stats:
   - The `xmeminfo' utility is now optional, and the output from the tool is
     now written to separate file (was previously appended to usage.csv). The
     error messages from xmeminfo written to 'xmeminfo.log' (was 'xres.log').
   - Disk usage output from `df' is now written to separate file (was
     previously appended to usage.csv).
   - Collect /sys/class/power_supply/ and /sys/class/backlight/ recursively.
   - Large disk reads and writes filtered through `sp-noncached' utility to
     avoid polluting kernel page cache with endurance data.
 * proc2csv:
   - Collect /proc/pid/wchan and /proc/pid/io.
   - The '-t' testing flag is removed.
 * endurance_report.py:
   - Report Upstart respawned jobs.
   - Disk space usage changes shown for all filesystems (instead of just / and
     /tmp).
 * syslog_parse.py:
   - Now uses configuration file for matching interesting lines from syslog.
     This utility now provides results in text format only.
   - The Python API that this file exports was changed.
 * endurance_plot:
   - Graphs now categorized to 'process' and 'system' groups.
   - New graphs: disk I/O graphs that show MB per round, ext4 filesystem
     writes per round, process counts in different wait channels, process
     counts in different non-sleep states, separate networking plots per
     network interface, /dev/pvrsrvkm mapping count.
   - Improvements to graphs: legends more readable, x-axis descriptions added
     that show round number plus date and time.
   - Switched to `pngcairo' gnuplot terminal.
   - Command line switch '-j' added for executing some processing in parallel.
 * recompress-endurance-measurements: New post-processing utility for
   re-compressing endurance data that was compressed with lzop.

sp-endurance (2.3), 27 Oct 2011:
 * extract-endurance-process-cgroups:
   - New utility that can be used to extract process Cgroup information from
     endurance data. Produces results in either text or HTML format.
 * endurance_report.py:

src/proc2csv.c
/*
 * A tiny '/proc information output' utility.
 *
 * Compile with: gcc -o proc2csv -W -Wall -Os -s proc2csv.c
 * 
 * This reads the PIDs of all processes at startup and then outputs
 * the status of those processes.  Kernel processes and processes which
 * have exited, are silently ignored.  Then some system info from
 * other /proc files is output.  All output is in CSV format.
 * 
 * Files that are processed from /proc:
 * - uptime
 * - loadavg
 * - meminfo
 * - vmstat
 * - sysvipc/msg
 * - sysvipc/sem
 * - sysvipc/shm
 * - sys/fs/file-nr
 * - PID/cmdline
 * - PID/fds/ (just the count of open FDs)
 * - PID/stat
 * - PID/status
 * - PID/wchan
 * - PID/io (optional)
 * 
 * NOTES:
 * - Originally this was a 'top' utility contributed by Eero
 *   to Busybox, but it is now severely re-factored & generalized,
 *   reading of many other proc files is added and output changed
 *   to CSV and to happen only once
 * - At startup this changes to /proc, all the reads are then
 *   relative to that
 * 
 * Copyright (C) 2003 by Eero Tamminen
 * Copyright (C) 2006,2007,2009,2011,2012 by Nokia Corporation
 * 
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * version 2 as published by the Free Software Foundation. 
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <dirent.h>
#include <string.h>
#include <sys/types.h>
#include <dirent.h>
#include <libgen.h>	/* basename() */
#include <assert.h>
#include <errno.h>

/* set to non-zero value to enable "-t" option which can be used
 * to test parsing of /proc contents copied from other machines.
 * 
 * disabled by default as it can have security implications when
 * proc2scv has been granted extra capabilities.
 */
#define PROC_TEST 0

/* field separator and what that is replaced with if it appears
 * in the input.
 */
#define CSV_SEPARATOR ','
#define SEPARATOR_REPLACEMENT '/'

/* process information taken from /proc,
 * The code takes into account how long the fields below are.
 */
typedef struct {
	int skip;	/* whether to skip this item */
	char cmd[128];	/* command line[read/show size] */
	char pid[6];	/* Pid */
} status_t;

/* return values for process types */
typedef enum {
	PID_OK,
	PID_SKIP,
	PID_EXITED
} pid_type_t;

/* whether to show /proc file fields or values */
enum {
	SHOW_FIELDS,
	SHOW_VALUES
};

static int ignore_user_errors = 0;

static inline void newline(void)  { fputc('\n', stdout); }

/* show the related error messages and exit unless
 * ignore_user_errors is set
 */
static void error_exit(const char *fun, const char *msg, const char *file)
{
	int error = errno;
	
	perror(fun);
	if (ignore_user_errors && error == EACCES) {
		fprintf(stderr, "  INFO: %s for '%s'\n", msg, file);
		return;
	} else {
		fprintf(stderr, "  ERROR: %s for '%s'\n", msg, file);
		exit(-1);
	}
}

/* converts contents of given file to CSV so that all whitespace and
 * control characters are compressed and replaced with CSV_SEPARATOR.
 * CSV_SEPARATOR chars in the input itself are replaced.
 * For example "foo    bar" -> "foo,bar"
 * 
 * Only count columns from 'start'th one are output.
 */
static void show_as_csv(const char *filename, int start, int count)
{
	char buffer[512], *buf, *value;
	int field, fields_out;
	FILE *fp;

	fp = fopen(filename, "r");
	if (!fp) {
		error_exit("show_as_csv()", "file open failed", filename);
		return;
	}
	
	while (fgets(buffer, sizeof(buffer), fp)) {
		buf = buffer;
		fields_out = 0;
		for (field = 0; fields_out < count; field++) {
			while (*buf && *buf <= ' ') {
				buf++;
			}
			if (*buf) {
				value = buf;
			} else {
				break;
			}
			while (*buf && *buf > ' ') {
				if (*buf == CSV_SEPARATOR) {
					*buf = SEPARATOR_REPLACEMENT;
				}
				buf++;
			}
			if (field < start) {
				continue;
			}
			if (*buf) {
				*buf++ = '\0';
			}
			if (fields_out) {
				fputc(CSV_SEPARATOR, stdout);
			} else {
				fields_out = 1;
			}
			fputs(value, stdout);
		}
		newline();
	}
	fclose(fp);
}


/* splits the given buffer into 'sep' separated key and value strings
 * which are stripped of white space & control chars from both ends.
 * CSV_SEPARATOR chars in the value part are replaced.
 * Returns zero on success and -1 on error.
 */
static int split_key_value(char *buf, char **key, char **value, char sep)
{
	assert(buf);
	while (*buf && *buf <= ' ') {
		buf++;
	}
	*key = buf;
	while (*buf && *buf != sep) {
		buf++;
	}
	if (*buf != sep) {
		*value = buf;
		return -1;
	}
	*buf++ = '\0';
	while (*buf && *buf <= ' ') {
		buf++;
	}
	*value = buf;
	while (*buf) {
		if (*buf == CSV_SEPARATOR) {
			*buf = SEPARATOR_REPLACEMENT;
		}
		buf++;
	}
	while (buf > *value) {
		buf--;
		if (*buf <= ' ') {
			*buf-- = '\0';
		} else {
			break;
		}
	}
	return 0;
}


/* split the given line at 'separator' char to key and value
 * and print either of them according to 'show'
 */
static void output_fields(FILE *fp, int show, char separator)
{
	char buf[512], *key, *value;
	int fields = 0;
	
	while (fgets(buf, sizeof(buf), fp)) {
		if (buf[0] <= ' ') {
			/* lines starting with whitespace don't have
			 * proper key/value pairs
			 */
			continue;
		}
		if (fields) {
			fputc(CSV_SEPARATOR, stdout);
		} else {
			fields = 1;
		}
		if (split_key_value(buf, &key, &value, separator) < 0) {
			fprintf(stderr, "ERROR: buffer '%s' didn't contain '%c'!\n", buf, separator);
			continue;
		}
		if (show == SHOW_FIELDS) {
			fputs(key, stdout);
		} else {
			fputs(value, stdout);
		}
	}
	if (show == SHOW_FIELDS) {
		fputs(":", stdout);
	}
	newline();
}


static void show_keyvalue_file(const char *filename, char separator)
{
	FILE *fp;

	fp = fopen(filename, "r");
	if (!fp) {
		error_exit("show_keyvalue_file()", "file open failed", filename);
		return;
	}
	output_fields(fp, SHOW_FIELDS, separator);
	rewind(fp);
	output_fields(fp, SHOW_VALUES, separator);
	fclose(fp);
}


static pid_type_t show_status(status_t *s, int show)
{
	char status[20];
	FILE *fp;

	if (s->skip) {
		return PID_SKIP;
	}
	/* read the process info from 'status' in PID dir */
	snprintf(status, sizeof(status), "%s/status", s->pid);
	fp = fopen(status, "r");
	if (!fp) {
		if (errno != ENOENT) {
			error_exit("show_status()",
				   "file open failed", status);
		}
		/* skip already exited processes */
		s->skip = 1;
		return PID_EXITED;
	}
	output_fields(fp, show, ':');
	fclose(fp);
	return PID_OK;
}


/* read process statuses */
static void show_statuses(int num, status_t *statuslist)
{
	int idx, exited = 0;
	status_t *s;
	
	/* output CSV header for status file fields... */
	for (s = statuslist, idx = 0; idx < num; idx++, s++) {
		if (show_status(s, SHOW_FIELDS) == PID_OK) {
			/* fields printed OK from one status file */
			break;
		}
	}
	/* ...and the status file field values */
	for (s = statuslist, idx = 0; idx < num; idx++, s++) {
		switch (show_status(s, SHOW_VALUES)) {
		case PID_EXITED:
			exited++;
			break;
		case PID_SKIP:
		case PID_OK:
			break;
		}
	}
	if (exited) {
		fprintf(stderr,
			"%d (more) processes had exited in the meanwhile.\n",
			exited);
	}
}


/* read /proc/pid/stat */
static void show_proc_pid_stat(int num, const status_t *statuslist)
{
	char stat[20];
	int i;
	const status_t *s;
	for (i=0; i < num; ++i) {
		s = &statuslist[i];
		if (s->skip) {
			continue;
		}
		snprintf(stat, sizeof(stat), "%s/stat", s->pid);
		show_as_csv(stat, 0, 128);
	}
}

static void show_proc_pid_wchan(int num, const status_t *statuslist)
{
	char buffer[256];
	FILE *fp;
	int i;
	for (i=0; i < num; ++i) {
		const status_t *s = &statuslist[i];
		if (s->skip)
			continue;
		snprintf(buffer, sizeof(buffer), "%s/wchan", s->pid);
		buffer[sizeof(buffer)-1] = 0;
		fp = fopen(buffer, "r");
		if (!fp) {
			error_exit("show_proc_pid_wchan()", "file open failed",
					buffer);
			continue;
		}
		if (fgets(buffer, sizeof(buffer), fp)) {
			buffer[sizeof(buffer)-1] = 0;
			if (strlen(buffer) > 0) {
				fprintf(stdout, "%s%c%s\n",
					s->pid, CSV_SEPARATOR, buffer);
			}
		}
		fclose(fp);
	}
}

static void show_proc_pid_io(int num, const status_t *statuslist)
{
	char buffer[20];
	FILE *fp;
	int i;
	int header = 0;
	for (i=0; i < num; ++i) {
		const status_t *s = &statuslist[i];
		if (s->skip)
			continue;
		snprintf(buffer, sizeof(buffer), "%s/io", s->pid);
		buffer[sizeof(buffer)-1] = 0;
		fp = fopen(buffer, "r");
		if (!fp) {
			/* Ignore, /proc/pid/io is not universally available. */
			continue;
		}
		if (!header) {
			fputs("\nPID,", stdout);
			output_fields(fp, SHOW_FIELDS, ':');
			rewind(fp);
		}
		header = 1;
		fprintf(stdout, "%s%c", s->pid, CSV_SEPARATOR);
		output_fields(fp, SHOW_VALUES, ':');
		fclose(fp);
	}
}

/* read fd counts for each process in statuslist */
static void show_fd_counts(int num, status_t *statuslist)
{
	int fds, idx, exited = 0;
	char fddir[20];
	status_t *s;
	DIR *dir;

	fputs("PID,FD count,Command line:\n", stdout);
	for (s = statuslist, idx = 0; idx < num; idx++, s++) {

		if (s->skip) {
			continue;
		}
		/* open the dir containing process FDs */
		snprintf(fddir, sizeof(fddir), "%s/fd", s->pid);
		dir = opendir(fddir);
		if (!dir) {
			if (errno != ENOENT) {
				error_exit("show_fd_counts()",
					   "directory open failed", fddir);
			}
			/* skip already exited processes */
			s->skip = 1;
			exited++;
			continue;
		}
		/* count files in the fd/ subdirectory */
		for (fds = 0; readdir(dir); fds++)
		  ;
		closedir(dir);

		/* ignore current and parent dir entries */
		fds -= 2;
		assert(fds >= 0);
		fprintf(stdout, "%s%c%d%c%s\n",
			s->pid, CSV_SEPARATOR, fds, CSV_SEPARATOR, s->cmd);
	}
	if (exited) {
		fprintf(stderr,
			"%d (more) processes had exited in the meanwhile.\n",
			exited);
	}
}


/* allocs statuslist and reads process command lines, frees namelist
 * (which was allocated by scandir()), returns filled statuslist or
 * NULL in case of error.  In case of an error, namelist may be only
 * half freed.
 */
static status_t *read_info(int num, struct dirent **namelist)
{	
	FILE *fp;
	struct dirent **n;
	status_t *statuslist, *s;
	int i, idx, count, exited = 0;
	char filename[20], *cmdline;
	
	/* allocate & zero status for each of the processes */
	statuslist = calloc(num, sizeof(status_t));
	if (!statuslist) {
		return NULL;
	}

	/* go through the processes */
	n = namelist;
	s = statuslist;
	for (idx = 0; idx < num; idx++, s++, n++) {

		/* copy PID string to status struct and free name */
		if (strlen((*n)->d_name) > sizeof(s->pid)-1) {
			fprintf(stderr, "PID '%s' too long\n", (*n)->d_name);
			free(statuslist);
			return NULL;
		}
		strncpy(s->pid, (*n)->d_name, sizeof(s->pid));
		s->pid[sizeof(s->pid)-1] = '\0';
		free((*n));
		*n = NULL;

		/* read the command line from 'cmdline' in PID dir */
		snprintf(filename, sizeof(filename), "%s/cmdline", s->pid);
		fp = fopen(filename, "r");
		if (!fp) {
			/* skip already exited processes */
			s->skip = 1;
			exited++;
			continue;
		}
		count = fread(s->cmd, 1, sizeof(s->cmd)-1, fp);
		fclose(fp);
		
		cmdline = s->cmd;
		for (i = 0; i < count-1; i++) {
			if (cmdline[i] < ' ') {
				cmdline[i] = ' ';
			}
		}
		cmdline[++i] = '\0';
	}
	free(namelist);
	if (exited) {
		fprintf(stderr,
			"%d processes had exited in the meanwhile.\n",
			exited);
	}
	return statuslist;
}


/* returns true for file names which are PID dirs
 * (i.e. start with number)
 */
static int filter_pids(const struct dirent *dir)
{
	status_t dummy;
	const char *name = dir->d_name;

	if (*name >= '0' && *name <= '9') {
		if (strlen(name) > sizeof(dummy.pid)-1) {
			fprintf(stderr, "PID name '%s' too long\n", name);
			return 0;
		}
		if (atoi(name) == getpid()) {
			/* ignore myself */
			return 0;
		}
		return 1;
	}
	return 0;
}


/* compares two directory entry names as numeric strings
 */
static int num_sort(const struct dirent **a, const struct dirent **b)
{
	int ia = atoi((*a)->d_name);
	int ib = atoi((*b)->d_name);

	if (ia == ib) {
		return 0;
	}
	/* NOTE: by switching the check, you change the process sort order */
	if (ia < ib) {
		return -1;
	} else {
		return 1;
	}
}


static void usage(const char *name)
{
#if PROC_TEST
	printf("\nusage: %s [-t|-p]\n\n", name);
#else
	printf("\nusage: %s [-p]\n\n", name);
#endif
	printf(
"First this reads all PIDs in /proc, then it will read their status\n"
"and some other system information and output that in CSV format to\n"
"the standard output.\n\n"
#if PROC_TEST
"With the '-t' (test) option, the 'proc' subdirectory in the current\n"
"directory is used instead of the system /proc directory.\n\n"
#endif
"With the '-p' option you can run this as normal user, as then all\n"
"permission denied errors are ignored.\n");
	exit(-1);
}


int main(int argc, char *argv[])
{
	status_t *statuslist;
	const char *proc = "/proc", *arg;
	struct dirent **namelist;
	int lines;

	/* which dir to use for "proc" directory */
	if (argc > 1) {
		arg = argv[1];
		if (argc == 2 && arg[0] == '-' && arg[1] && !arg[2]) {
			switch (arg[1]) {
#if PROC_TEST
			case 't':
				proc = "proc";
				break;
#endif
			case 'p':
				ignore_user_errors = 1;
				break;
			default:
				usage(argv[0]);
			}
		} else {
			usage(argv[0]);
		}
	}
	
	/* change to proc */
	if (chdir(proc) < 0) {
		perror("chdir('proc')");
		return -1;
	}

	/* show some system information */
	fputs("\nUptime,Idletime (secs):\n", stdout);
	show_as_csv("uptime", 0, 2);
	fputs("\nLoadavg 1min,5min,15min,Running/all,Last PID:\n", stdout);
	show_as_csv("loadavg", 0, 5);

	/* memory usage and vmstat limits */
	newline();
	show_keyvalue_file("meminfo", ':');
	newline();
	show_keyvalue_file("vmstat", ' ');
	/* compatibility support for old Maemo sp-endurance-postproc */
	fputs("\nlowmem_maemo,dummy2,dummy3:\n0,0,0\n", stdout);

	/* sysV IPC memory usage */
	fputs("\nMessage queues:\n", stdout);
	show_as_csv("sysvipc/msg", 2, 6);
	fputs("\nSemaphore arrays:\n", stdout);
	show_as_csv("sysvipc/sem", 2, 4);
	fputs("\nShared memory segments:\n", stdout);
	show_as_csv("sysvipc/shm", 2, 6);

	/* do this before scanning PIDs so that count is not disturbed */
	fputs("\nAllocated FDs,Freed FDs,Max FDs:\n", stdout);
	show_as_csv("sys/fs/file-nr", 0, 3);
	
	/* read process IDs for all the processes from the procfs */
	lines = scandir(".", &namelist, filter_pids, num_sort);
	if (lines < 0) {
		perror("scandir('proc')");
		return -1;
	}
	if (!lines) {
		fprintf(stderr, "No /proc/PID/ entries\n");
		return -1;
	}

	/* read command line for each of the processes */
	statuslist = read_info(lines, namelist);
	if (!statuslist) {
		fprintf(stderr, "Error in reading proccesses information");
		return -1;
	}

	/* show how many file descriptors each process is using */
	newline();
	show_fd_counts(lines, statuslist);
	
	/* read and show status for each of the processes */
	newline();
	show_statuses(lines, statuslist);

	/* show /proc/pid/stat for each process */
	fputs("\nProcess status:\n", stdout);
	show_proc_pid_stat(lines, statuslist);
	
	fputs("\nPID,wchan:\n", stdout);
	show_proc_pid_wchan(lines, statuslist);

	show_proc_pid_io(lines, statuslist);

	free(statuslist);
	return 0;
}
